### DISCUSSION TITLE
Underscore in interface member name

### DISCUSSION THREAD

**Discussion by Kr00l** [twinbasic/twinbasic#862] [General]
I know that this limitation exists in VB6. I just want to ask what may be the reason for this. Of course tB offers a solution for this.
However, only in the legacy way of doing this is not possible in tB. For a reason ?

Starting point is a class with a member that contains an underscore.

```
Class FooBarClass
    
    Public Function Foo_Bar() As String
        Foo_Bar = "Foo Bar"
    End Function
    
End Class
```

Now the class that implements the interface. Using the new tB style of doing things.

```
Class MyClass
    Implements FooBarClass

    Private Function Foo_Bar() As String Implements NewProject.FooBarClass.Foo_Bar
    	Foo_Bar = "Bar Foo?"
    End Function
    
End Class
```

Now, testing it. Works!
```
Public Sub Main()
    Dim i As New MyClass
    Dim x As FooBarClass
    Set x = i
    Debug.Print x.Foo_Bar
End Sub
```

The legacy of the MyClass does not work. Also not in VB6. But, must it be like this? Or could tB allow it anyway? It wouldn't break compatibility ?

```
Class MyClass
    Implements FooBarClass
    
    Private Function FooBarClass_Foo_Bar() As String
        FooBarClass_Foo_Bar = "Bar Foo?"
    End Function
    
End Class
```
![image](https://user-images.githubusercontent.com/33756666/169535199-4f090a0f-8f7e-43e7-8f28-d4de2df5d2ba.png)

**Comment by bclothier** [comment]
[MS-VBAL Section 5.2.4.2 says no](https://docs.microsoft.com/en-us/openspecs/microsoft_general_purpose_programming_languages/ms-vbal/da526020-9b41-44a6-a5f3-47a7ac255a9e).

> A class can’t be used as an interface class if the names of any of its public variable or method methods contain an underscore character (Unicode u+005F).

I think for backward compatibility it's best to comply with the restriction as specified. 

Also, consider the scenario where we provide a class that may be then implemented in the VBA host or in VB6.

**Reply by Kr00l** [reply]
Ok, agree. Should the tB style of implementing then respect this "rule" or not ?
IMO if there is no technical reason, it should not follow that "rule" as backward compatibility is no issue in that.

**Reply by bclothier** [reply]
I agree that if there is no restrictions, one can use the new style to work around the underscore but we have to keep in mind that such implementation can cause problems downstream especially if those members are exposed to other consumers outside the tB environment. As long it's internal to the tB codebase, I don't see a problem. It's the public members I'm less sure about.

We could also have a simple compiler warning when such member is exposed to call attention to this. I don't think many people know about this particular restriction.

**Comment by wqweto** [comment]
This compiles in VB6
```
Option Explicit
Implements Foo_Bar

Private Sub Foo_Bar_Test()

End Sub
```

**Reply by bclothier** [reply]
This works because it's the class name that contains an underscore. The way VBx resolves the implementation/handlers is to split the member on the last underscore character, so it becomes class `Foo_Bar` and method `Test`. If the class was `Foo` and the method `Bar_Test`, that would fail to match the last underscore split behavior.

**Comment by WaynePhillipsEA** [comment]
What about for `Sub My_Silly_Interface_Name_My_Even_Sillier_Method_Name()`?  

I think we should stick to the current handling, and promote the Implements/Handles syntax for these cases.

**Reply by bclothier** [reply]
Under VBx, I'm pretty sure it'll try to resolve as:
Class:`My_Silly_Interface_Name_My_Even_Sillier_Method` (doesn't exist)
Method:`Name` 

IDE is confused:
![image](https://user-images.githubusercontent.com/2367644/169608399-9c1230b2-467c-4c95-8fb8-0917b0d38b81.png)

![image](https://user-images.githubusercontent.com/2367644/169608494-5ba163ef-6d36-4c50-8097-48e8e291f0b8.png)

![image](https://user-images.githubusercontent.com/2367644/169608732-29044fae-3f29-451a-9566-8aea8709e8f8.png)

**Reply by Kr00l** [reply]
When tB uses the new style implement method name with underscore and exposes this? Can it cause problems due to that "rule"? (What bclothier pointed out)
